cpp/src/groupby/hash/groupby.cu

```diff
@@ -68,8 +81,248 @@ bool constexpr is_hash_aggregation(aggregation::Kind t) {
+      columns.push_back(request.values);
+      col_ids.push_back(i);
+    };
+
+    std::for_each(agg_v.begin(), agg_v.end(),
+      [&columns, &agg_kinds, &request, &col_ids, insert_agg] 
+      (std::unique_ptr<aggregation> const& agg) {
+
+        if (is_hash_aggregation(agg->kind)) {
+          if (is_fixed_width(request.values.type())) {
+            insert_agg(agg->kind);
+          } else if (request.values.type().id() == type_id::STRING) {
+            // For string type, only ARGMIN, ARGMAX, MIN, and MAX are supported
+            if (agg->kind == aggregation::ARGMIN or 
+                agg->kind == aggregation::ARGMAX) {
+              insert_agg(agg->kind);
+            } else if (agg->kind == aggregation::MIN) {
+              insert_agg(aggregation::ARGMIN);
+            } else if (agg->kind == aggregation::MAX) {
+              insert_agg(aggregation::ARGMAX);
+            }
+          }
+        }
+      });
+  }
+  return std::make_tuple(table_view(columns), 
+                         std::move(agg_kinds), std::move(col_ids));
+}
+
+/**
+ * @brief Gather sparse results into dense using @p gather_map and add to 
+ * @p dense_cache
+ */
+void sparse_to_dense_results(
+    std::vector<aggregation_request> const& requests,
+    experimental::detail::result_cache const& sparse_results,
+    experimental::detail::result_cache& dense_results,
+    rmm::device_vector<size_type> const& gather_map,
+    cudaStream_t stream,
+    rmm::mr::device_memory_resource* mr)
+{
+  for (size_t i = 0; i < requests.size(); i++) {
+    auto const& agg_v = requests[i].aggregations;
+    auto const& col = requests[i].values;
+
+    // Given an aggregation, this will get the result from sparse_results and 
+    // convert and return dense, compacted result
+    auto to_dense_agg_result =
+    [&sparse_results, &gather_map, i, mr, stream]
+    (auto const& agg) {
+      auto s = sparse_results.get_result(i, agg);
+      auto dense_result_table = 
+        experimental::detail::gather(
+          table_view({s}),
+          gather_map.begin(),
+          gather_map.end(),
+          false, false, false, mr, stream);
+      auto dense_result = std::move(dense_result_table->release()[0]);
+      return dense_result;
+    };
+
+    // Enables conversion of ARGMIN/ARGMAX into MIN/MAX
+    auto transformed_result =
+    [&col, to_dense_agg_result, mr, stream]
+    (auto const& agg_kind) {
+      auto tranformed_agg = std::make_unique<aggregation>(agg_kind);
+      auto argmax_result = to_dense_agg_result(tranformed_agg);
+      auto transformed_result = experimental::detail::gather(
+        table_view({col}), *argmax_result, false, false, false, mr, stream);
+      return std::move(transformed_result->release()[0]);
+    };
+
+    std::for_each(agg_v.begin(), agg_v.end(),
+      [&sparse_results, &dense_results, to_dense_agg_result, transformed_result,
+       &col, i]
+      (auto const& agg) {
+        if (col.type().id() == type_id::STRING and
```

Comment:
This is another place where it would be cleaner and more maintainable to define the mapping of compound aggs to their primitive components (and back again) in some kind of trait.